### ISSUE TITLE
Missing environment cube map when playing Freelancer (2003) on modern OS/DirectX 8 version

### ISSUE THREAD

**Issue by BC46** [elishacloud/dxwrapper#136]
In the game Freelancer (2003) the space ships are supposed to have a somewhat-shiny front window. This shininess is achieved in the game by applying an environment cube map on top of all ship windows. This is what the cube map looks like:
![image](https://user-images.githubusercontent.com/45081813/227371043-9ee2899d-3af5-43c3-9409-6914bad40dd1.png)

Unfortunately after some DirectX 8 update this particular cube map stopped working completely which caused all front ship windows to lose their shininess. The exact update where this started to occur is unknown. Though it's certain that it happened a very long time ago since this issue is present on Windows 7 and even on the latest Windows XP service pack.

The only known way of getting the cube map to work on modern operating systems is to launch the game using dgVoodoo. It would be nice if this could also be fixed in DxWrapper, or possibly d3d8to9.

Windows 10:
![image](https://user-images.githubusercontent.com/45081813/227374551-9439d1ad-eead-4cdc-99ef-3ee19a8e7a2d.png)

dgVoodoo:
![image](https://user-images.githubusercontent.com/45081813/227374565-7954cf62-912d-4c64-a192-a805af96e24a.png)

**Comment by BC46** [comment]
I've updated the content of this issue because one major part of the mentioned problem has already been fixed during this discussion: https://github.com/elishacloud/dxwrapper/issues/134.

**Comment by elishacloud** [comment]
Do you have an older system setup where this works?  If so can you send me a d3d9 Microsoft PIX output file of that frame working correctly?

**Comment by BC46** [comment]
I happen to have an ancient Windows XP laptop lying around. Might be a while before I can get that thing running again...

**Comment by BC46** [comment]
Managed to get the game running on the laptop. There's two problems though. First of all, for some reason even on this system the cube map does not work. I tried installing the DirectX version that was shipped with the game, but to no avail.

The second problem is that I can't get d3d8to9 to work on the laptop. It says a file named "MSVCP140.dll" is missing. Installing the VC++ 12.0 redistributable package did not do the trick unfortunately.

**Comment by mirh** [comment]
140 is visual studio 2015
(hoping that crosire used the XP version of the toolchain)
-------- Messaggio originale --------
Il 25 mar 2023, 19:58, C.S. Melis ha scritto:

**Comment by elishacloud** [comment]
Here is a version complied for Windows XP with the runtimes built-in:  [d3d8.zip](https://github.com/elishacloud/dxwrapper/files/11070626/d3d8.zip)

**Comment by BC46** [comment]
> 140 is visual studio 2015

Ah, I didn't know the number at the end represents the VC++ version.

> Here is a version complied for Windows XP with the runtimes built-in: [d3d8.zip](https://github.com/elishacloud/dxwrapper/files/11070626/d3d8.zip)

Thank you for this build! Running the game with it no longer showed the error message, but it did crash right after launch.
ModName: `d3d8.dll`
Offset: `00013809`

I tried installing a VC++ 14 Redistributable package that was still supported on XP. After that the original d3d8to9 dll did load, but got the same crash as above, but this time with a different crash offset: `00013f53`.

**Comment by elishacloud** [comment]
This issue might be related to issue #75

**Comment by elishacloud** [comment]
Ok, here is a quick test build to see if this will fix the missing items:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/11215934/dxwrapper.zip)

**Comment by BC46** [comment]
It seems that didn't do the trick unfortunately: 
![image](https://user-images.githubusercontent.com/45081813/231586255-fca11442-a3dc-407d-8ffe-d982654e4d97.png)

**Comment by elishacloud** [comment]
I see.  It was worth a try.  Thanks for checking.

**Comment by BC46** [comment]
For the past few days I had another look at this issue and I managed to find a solution for modern DirectX versions (within Freelancer).

The glass of the space ship is a surface with its own material which is specified in the game files. This glass material solely consists of a diffuse color and opacity; it has no texture assigned.

To make the environment cube map render on the glass surface, all you have to do is assign a "dummy" texture to the glass' material via the game files. I used a small solid white 32-bit TGA texture: 
![Image](https://github.com/user-attachments/assets/028b6e2a-5690-4a28-81d5-bff57d06524e)

Before and after screenshots:

![Image](https://github.com/user-attachments/assets/9c2f118f-1cb5-42ae-865f-cafbea7ef67a)

![Image](https://github.com/user-attachments/assets/2737f218-77c3-4f6c-969d-040b59e81a6a)


Long story short, it appears that in modern D3D8 you can't apply an environment cube map to a textureless surface. However, it seems this isn't an issue in legacy D3D8. I don't know if you'd like to implement a general solution for this in dxwrapper since it's technically a compatibility issue between legacy D3D8 and modern D3D8. If this lies outside the project's scope, then this issue can be closed as far as I'm concerned.

**Comment by elishacloud** [comment]
> To make the environment cube map render on the glass surface, all you have to do is assign a "dummy" texture to the glass' material via the game files.

Great insight.  I don't have any real way to test this, but I put a fix in that tries to check if an environment cube map is being used with a textureless surface and then put in a dummy texture to allow it to work.

See if this update helps:  [dxwrapper.zip](https://github.com/user-attachments/files/18613415/dxwrapper.zip)

**Comment by BC46** [comment]
> Great insight. 

Thanks! This has actually been on my list for almost 4 years now. I feel very relieved to have finally found a proper solution for it. 😃

> See if this update helps: [dxwrapper.zip](https://github.com/user-attachments/files/18613415/dxwrapper.zip)

![Image](https://github.com/user-attachments/assets/f65ff2c9-0045-4ef6-91e2-2db84f4122b3)

I tested this build with an unmodified copy of the game but the env cube map doesn't render. What kind of dummy texture did you put in?

**Comment by mirh** [comment]
What you call "modern d3d8" is probably microsoft's new big brain idea of shimming it to d3d9 (as I think we had already established in #134?). And bugs means it's bugged, there isn't much to argue.
It is one of the problems dxwrapper is supposed to handle, but maybe you could also check over with d3d8to9 and dxvk.

**Comment by elishacloud** [comment]
> Long story short, it appears that in modern D3D8 you can't apply an environment cube map to a textureless surface.

What makes you think this is an environment cube map?  Also, what is your definition of an environment cube map?

> maybe you could also check over with d3d8to9 and dxvk.

Just out of curiosity, did you check with dxvk instead of d3d8to9?  Does the same issue happen with both d3d8to9 and dxvk?

> What kind of dummy texture did you put in?

I just created a new blank 64x64 2D texture and set it when I detected that there was an environment cube map with no texture.  However, I may not have been checking for all cases where an environment cube map is used.  Also, I maybe should have been using a texture cube not a 2D texture.

Here is another text build: [dxwrapper.zip](https://github.com/user-attachments/files/18630398/dxwrapper.zip)

**Comment by BC46** [comment]
> What you call "modern d3d8" is probably microsoft's new big brain idea of shimming it to d3d9 (as I think we had already established in https://github.com/elishacloud/dxwrapper/issues/134?).

Yes, the lighting bug which we discussed has been present since Windows 10 version 2004 which rolled out somewhere in May 2020. Presumably that's when (a part of) the D3D9 conversion started happening. Though as I explained in https://github.com/elishacloud/dxwrapper/issues/136#issuecomment-1483897252, the glass cube map was missing even on my old Windows XP laptop which hasn't seen a Windows update in over 15 years. Therefore, the missing glass cube map problem has existed for much longer than the lighting bug.

> What makes you think this is an environment cube map?

The glass texture is stored inside a file called `envmapglass.txm`. As for the cube part, the texture was exported in DDS format as a Cube Map (horizontal cross layout). Hence you can see all six faces of the cube as shown in the image I provided in the issue description. Moreover, I intercepted the `CCubeMap::GetCubeMapSurface` calls to `d3d8.dll` and I could see the game calling this function explicitly for the glass env cube map (a pointer to the string `envmapglass` was present on the stack).

> Also, what is your definition of an environment cube map?

I'll give an informal definition. Applying an environment cube map to a surface allows the appearance of that surface to change depending on how the camera is positioned relative to that surface. The environment cube map consists of six textures which each represent a side of the cube. As the camera moves around the surface, the renderer will map different faces of the cube to that surface. Freelancer uses this technique as a cheap way to make metally and glassy surfaces appear reflective. The metal reflections are done using a different env cube map which is in a file named `envmapbasic.txm`. 

Below is a video demonstrating how the surfaces change when I move the camera around the ship. Notice how the metal surfaces appear to be shiny and how the glass becomes dark when the camera moves up. This is all done using environment cube maps.

https://github.com/user-attachments/assets/3323958a-5501-4ccb-a9b1-b8e83eeca9f8

In the video below I turned off all env cube maps. Notice how this makes the surfaces of the ship look dull.

https://github.com/user-attachments/assets/fa507c44-e555-4726-a9e6-88adb1c96fa2

> Does it have to be a white texture or does it work with a black one?

If I make the dummy texture black, then the glass turns black too. So yes I think it has to be a solid white texture.

![Image](https://github.com/user-attachments/assets/d68bdbb6-1446-40b3-8447-033f47b0ace6)

> Just out of curiosity, did you check with dxvk instead of d3d8to9? Does the same issue happen with both d3d8to9 and dxvk?

I tested both before but did it again with the latest releases just to be sure. Yes, the issue happens on both. As far as I know, dgVoodoo is currently the only D3D8 API implementation that fixes the issue and is compatible with modern systems.

> Here is another test build: [dxwrapper.zip](https://github.com/user-attachments/files/18630469/dxwrapper.zip)

This one seems to definitely have an effect, but it makes the glass entirely invisible/transparent. Also some in-game HUD elements no longer render, either. Perhaps the dummy texture needs to have a solid color like in my example?

![Image](https://github.com/user-attachments/assets/67b44575-6cd6-4f2f-9552-4ca90328ae96)

**Comment by elishacloud** [comment]
@BC46, thanks for all your comments.

> This one seems to definitely have an effect, but it makes the glass entirely invisible/transparent. Also some in-game HUD elements no longer render, either. Perhaps the dummy texture needs to have a solid color like in my example?

This is good.  It means we are in the right direction.  Currently I just create a texture and don't set any color.  This creates a full black texture.  Let me try setting a texture color.

**Comment by BC46** [comment]
Interesting. The glass is now black, but so are the in-game HUD elements which were previously transparent.

![Image](https://github.com/user-attachments/assets/35f5ccef-f460-4b06-a7ba-745bba1e7245)

**Comment by elishacloud** [comment]
Thanks for the quick testing.  Looks like ColorFill() did not work, as the textures are still black.  Try this one:  [dxwrapper.zip](https://github.com/user-attachments/files/18698536/dxwrapper.zip)

**Comment by BC46** [comment]
> Thanks for the quick testing. 

Anytime!

> Try this one: [dxwrapper.zip](https://github.com/user-attachments/files/18698536/dxwrapper.zip)

![Image](https://github.com/user-attachments/assets/b863f38c-5994-40d2-9382-c4659051e838)

Success! The glass env map renders correctly with this build. Also the HUD elements which were previously black now show up intended. Only strange thing I noticed is that whenever I Alt-Tab and then Tab back into the game, it crashes.

![Image](https://github.com/user-attachments/assets/29e426c5-ab64-4331-812b-69cc66182622)

dxwrapper wants to zero 0xFFE9A7FC bytes at address 0x607F000.

**Comment by elishacloud** [comment]
> dxwrapper wants to zero 0xFFE9A7FC bytes at address 0x607F000.

Oops.  That was just a typo.  I will fix this and then give you one last build to test.

**Comment by elishacloud** [comment]
Here is the testing build: [dxwrapper.zip](https://github.com/user-attachments/files/18699092/dxwrapper.zip)

I added some options below:

```ini
[Testing]
DDCubeShaderFix            = 1
DDCubeTextureMapFix        = 1
DDCubeEnvironmentMapFix    = 1
DDCubeTransformMapFix      = 1
```

Can you test this and tell me which ones of these options need to be enabled for this fix to work?

**Comment by BC46** [comment]
Yes I can. With this config the env cube map shows up:

```ini
[Testing]
DDCubeShaderFix            = 0
DDCubeTextureMapFix        = 1
DDCubeEnvironmentMapFix    = 0
DDCubeTransformMapFix      = 0
```

But this one also works:
```ini
[Testing]
DDCubeShaderFix            = 0
DDCubeTextureMapFix        = 0
DDCubeEnvironmentMapFix    = 0
DDCubeTransformMapFix      = 1
```

And this config has no effect (env cube map doesn't show up):
```ini
[Testing]
DDCubeShaderFix            = 1
DDCubeTextureMapFix        = 0
DDCubeEnvironmentMapFix    = 1
DDCubeTransformMapFix      = 0
```

So it seems that if either `DDCubeTextureMapFix` or `DDCubeTransformMapFix` is enabled, then the env cube map works. Though `DDCubeShaderFix` and `DDCubeEnvironmentMapFix` do not help to make the env cube map work.

**Comment by elishacloud** [comment]
> So it seems that if either `DDCubeTextureMapFix` or `DDCubeTransformMapFix` is enabled, then the env cube map works. Though `DDCubeShaderFix` and `DDCubeEnvironmentMapFix` do not help to make the env cube map work.

Thanks!  That is not what I expected.  However, it is very useful.

Here is the final build.  If this works then I will check-in the code:  [dxwrapper.zip](https://github.com/user-attachments/files/18709472/dxwrapper.zip)

**Comment by BC46** [comment]
I tested it and the env cube map renders correctly with this build.

**Comment by elishacloud** [comment]
Thanks @BC46 for all your testing here!

**Comment by Squall-Leonhart** [comment]
> What you call "modern d3d8" is probably microsoft's new big brain idea of shimming it to d3d9 (as I think we had already established in [#134](https://github.com/elishacloud/dxwrapper/issues/134)?). And bugs means it's bugged, there isn't much to argue.

theres no shimming, theres just a fair amount of shared code between them.

**Comment by elishacloud** [comment]
@BC46, I just put a check-in to disable `EnvironmentCubeMapFix` by default (see #379).  This means, starting with the next release it will require you to enable this option to fix Freelancer.  If you get a chance to to test this update it would be appreciated.  You can download the build from [here](https://github.com/elishacloud/dxwrapper/actions/runs/14600104718).

**Comment by BC46** [comment]
Not a problem. I already fixed the issue in my project by adding the dummy texture directly to the game files. This way the environment cube map works without the use of DirectX-related compatibility fixes, too.

I tested the build with the options `D3d8to9` and `EnvironmentCubeMapFix` both enabled and the fix still works.

**Comment by elishacloud** [comment]
> I tested the build with the options `D3d8to9` and `EnvironmentCubeMapFix` both enabled and the fix still works.

Great!  Thanks for checking.
